shared/test-helpers/ganache-cli.sh

```diff
@@ -1,50 +1,64 @@
 #!/usr/bin/env bash
 
 # Exit script as soon as a command fails.
-set -o errexit
+set -o errexit -o pipefail
 
-if [ "$SOLIDITY_COVERAGE" = true ]; then
-  testrpc_port=8555
-else
-  testrpc_port=8545
-fi
+# Executes cleanup function at script exit.
+trap cleanup EXIT
 
-testrpc_running() {
-  nc -z localhost "$testrpc_port"
+cleanup() {
+  # Kill the RPC instance that we started (if we started one and if it's still running).
+  if [ -n "$rpc_pid" ] && ps -p $rpc_pid > /dev/null; then
+    kill -9 $rpc_pid
+  fi
 }
 
-start_testrpc() {
-  echo "Starting our own testrpc instance at port $testrpc_port"
-  if [ "$SOLIDITY_COVERAGE" = true ]; then
-    npx testrpc-sc -i 16 --gasLimit 0xfffffffffff --port "$testrpc_port"  > /dev/null &
-  else
-    npx ganache-cli -i 15 --gasLimit 50000000 --port "$testrpc_port" > /dev/null &
-  fi
+setup_coverage_variables() {
+  PORT=${PORT-8555}
+  BALANCE=${BALANCE-100000}
+  GAS_LIMIT=${GAS_LIMIT-0xfffffffffff}
+  NETWORK_ID=${NETWORK_ID-16}
+}
 
-  testrpc_pid=$!
+setup_testing_variables() {
+  PORT=${PORT-8545}
+  BALANCE=${BALANCE-100000}
+  GAS_LIMIT=${GAS_LIMIT-8000000}
+  NETWORK_ID=${NETWORK_ID-15}
 }
 
-if testrpc_running; then
```

Comment:
Aren't we killing previous instances of testrpc anymore? Is this on purpose?

Reply:
It was on purpose: the idea was to avoid relying on `nc` in order to kill a process that is running on the port we want to use. Instead I'm adding adding a `cleanup` function with `trap` to ensure we always end the rpc process we have started. If there is a foreign process running on that port we shouldn't kill it, it's expected to fail rather than taking over a port while killing the process that is using it.

Reply:
Makes total sense!

Reply:
@facuspagnuolo How does ganache error if it can't bind to the requested port? If it's not very good, would it still make sense to test it before trying to run ganache on the same port so the error message is more human-readable?

Reply:
I will fail with a classic `EADDRINUSE` error, see:
```
Ganache CLI v6.4.2 (ganache-core: 2.5.4)
listen EADDRINUSE 127.0.0.1:8545
```

Reply:
That looks pretty good 😄